wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -245,6 +246,74 @@ public void stop(String workspaceId) throws NotFoundException, ServerException,
     }
 
     /**
+     * Adds machine into running workspace.
+     * This method do not touch workspace configuration.
+     * Just adds machine to workspace runtime and destroy it on workspace stop.
+     *
+     * @param machine
+     *         machine to add to specified runtime
+     * @throws NotFoundException
+     *         when workspace with specified id not running or not exists
+     * @throws ServerException
+     *         when application server is stopping
+     * @throws ConflictException
+     *         when workspace is not running
+     */
+    public void addMachine(MachineImpl machine) throws NotFoundException, ServerException, ConflictException {
+        ensurePreDestroyIsNotExecuted();
+
+        String workspaceId = machine.getWorkspaceId();
+
+        rwLock.writeLock().lock();
+        try {
+            ensurePreDestroyIsNotExecuted();
+            final RuntimeDescriptor descriptor = descriptors.get(workspaceId);
+            if (descriptor == null) {
+                throw new NotFoundException("Workspace with id '" + workspaceId + "' is not running.");
+            }
+            if (descriptor.getRuntimeStatus() != WorkspaceStatus.RUNNING) {
+                throw new ConflictException("Cannot add machine " + machine.getId() + " to not running workspace.");
+            }
+
+            descriptor.getRuntime().getMachines().add(machine);
+        } finally {
+            rwLock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * Removes machine from running workspace.
+     * This method do not touch workspace configuration and is opposite to {@link #addMachine(MachineImpl)}
+     * Does nothing if remove nonexistent machine
+     *
+     * @param machine
+     *         machine to remove from specified runtime
+     * @throws NotFoundException
+     *         when workspace with specified id not running or not exists
+     */
+    public void removeMachine(MachineImpl machine) throws NotFoundException {
+        String workspaceId = machine.getWorkspaceId();
+
+        rwLock.writeLock().lock();
+        try {
+            final RuntimeDescriptor descriptor = descriptors.get(workspaceId);
+            if (descriptor == null) {
+                throw new NotFoundException("Workspace with id '" + workspaceId + "' is not running.");
+            }
+
+            List<MachineImpl> machines = descriptor.getRuntime().getMachines();
+            for (int i = 0; i < machines.size(); i++) {
+                if (machine.getId().equals(machines.get(i).getId())) {
+                    machines.remove(i);
```

Comment:
Okay let me explain you my point of view:

```java
for (int i = 0; i < machines.size(); i++) {
    if (machine.getId().equals(machines.get(i).getId())) {
        machines.remove(i);
        return;
    }
}
```

From my comment: 
For ArrayList in the worst case(when the target element is the last one) get method will be invoked N times where N is the size of the list, so as the get from the ArrayList is O(1), then the search complexity is O(N).

For LinkedList  in the worst case(when the target element is the last one) get method will be performed N times as well, but as you mentioned get for LinkedList is O(N), so for the list from 4 elemenets we will have (1 + 2 + 3 + 4) operations, which is arithmetic progression with *1* as a step, so to get the complexity we use the arithmetic progression formula which is  `N * (A1 + An) / 2`, so in our case it can be converted to the `N(1+N)/2` which is actually exactly that count of operations we need, but yes according to the O notation it is O(n^2).

> P.S. Still can't figure out where did you get such formula: 1/2 * N * (1 + N). Does that correspond to some average-case complexity measurement?
I didn't say anything about average case complexity measurement.
There are were 2 examples for ArrayList and for LinkedList.